sympy/polys/matrices/domainmatrix.py

```diff
@@ -1692,6 +1750,582 @@ def det(self):
+            in the ground domain (``exquo``). This is also suitable for most
+            domains. This method can be used with overdetermined systems (more
+            equations than unknowns) but not underdetermined systems as a
+            unique solution is sought.
+
+        Returns
+        =======
+
+        (xnum, xden) : (DomainMatrix, DomainElement)
+            The solution of the equation $Ax = b$ as a pair consisting of an
+            ``n x m`` matrix numerator ``xnum`` and a scalar denominator
+            ``xden``.
+
+        The solution $x$ is given by ``x = xnum / xden``. The division free
+        invariant is ``A * xnum == xden * b``. If $A$ is square then the
+        denominator ``xden`` will be a divisor of the determinant $det(A)$.
+
+        Raises
+        ======
+
+        DMNonInvertibleMatrixError
+            If the system $Ax = b$ does not have a unique solution.
+
+        See Also
+        ========
+
+        solve_den_charpoly
+        solve_den_rref
+        inv_den
+        """
+        m, n = self.shape
+        bm, bn = b.shape
+
+        if m != bm:
+            raise DMShapeError("Matrix equation shape mismatch.")
+
+        if method is None:
+            method = 'rref'
+        elif method == 'charpoly' and m != n:
+            raise DMNonSquareMatrixError("method='charpoly' requires a square matrix.")
+
+        if method == 'charpoly':
+            xnum, xden = self.solve_den_charpoly(b)
+        elif method == 'rref':
+            xnum, xden = self.solve_den_rref(b)
+        else:
+            raise DMBadInputError("method should be 'rref' or 'charpoly'")
+
+        return xnum, xden
+
+    def solve_den_rref(self, b):
+        """
+        Solve matrix equation $Ax = b$ using fraction-free RREF
+
+        Solves the matrix equation $Ax = b$ for $x$ and returns the solution
+        as a numerator/denominator pair.
+
+        Examples
+        ========
+
+        >>> from sympy import ZZ
+        >>> from sympy.polys.matrices import DM
+        >>> A = DM([[ZZ(1), ZZ(2)], [ZZ(3), ZZ(4)]], ZZ)
+        >>> b = DM([[ZZ(5)], [ZZ(6)]], ZZ)
+        >>> xnum, xden = A.solve_den_rref(b)
+        >>> xden
+        -2
+        >>> xnum
+        DomainMatrix([[8], [-9]], (2, 1), ZZ)
+        >>> A * xnum == xden * b
+        True
+
+        See Also
+        ========
+
+        solve_den
+        solve_den_charpoly
+        """
+        A = self
+        m, n = A.shape
+        bm, bn = b.shape
+
+        if m != bm:
+            raise DMShapeError("Matrix equation shape mismatch.")
+
+        if m < n:
+            raise DMShapeError("Underdetermined matrix equation.")
+
+        Aaug = A.hstack(b)
+        Aaug_rref, denom, pivots = Aaug.rref_den()
+
+        if len(pivots) < n:
+            raise DMNonInvertibleMatrixError("Non-unique solution.")
+
+        # XXX: What if some pivots are after the nth column?
+        xnum = Aaug_rref[:n, n:]
```

Comment:
Has this xxx-todo been resolved?

Reply:
No, I need to fix it:
```python
In [515]: X = Matrix([[1, 2, 3], [4, 5, 6], [7, 8, 9]]).to_DM()

In [516]: b = Matrix([1, 2, 2]).to_DM()

In [517]: X.solve_den_rref(b)
Out[517]: (DomainMatrix({2: {0: 3}}, (3, 1), ZZ), mpz(3))

In [519]: X.hstack(b).rref_den()[0].to_Matrix()
Out[519]: 
⎡3  0  -3  0⎤
⎢           ⎥
⎢0  3  6   0⎥
⎢           ⎥
⎣0  0  0   3⎦
```
Here there is a pivot but it is in the rhs of the augmented matrix. I think we should change the rref routines so that they can distinguish between the lhs and rhs of an augmented matrix. A quick fix would just be to change the pivot check above to check that all pivots are in the expected range but in the event that the system does not have a unique solution it would be inefficient to continue eliminating everything on the rhs after we have already established that we cannot return a solution. In this case for example elimination should be stopped after 3 columns.

Reply:
This is fixed now